doc/developer/design/20240411_graceful_reconfiguration_of_managed_clusters_v1.md

```diff
@@ -0,0 +1,302 @@
+We will need to create a new table `mz_internal.mz_pending_cluster_replicas`
+
+| column     |  type |
+| ---------- | ----- |
+| replica_id | str   |
+| cluster_id | str   |
+| timeout    | int   |
+| wait_type  | str   |
+| background | bool  |
+| start_time | int   |
+| finished   | bool  |
+| continue_on_timeout | bool |
+
+This table will be responsible for holding the state of the graceful replication task.
+Each reconfiguration replica will have an entry in `mz_epnding_cluster_replicas`.
+
+
+#### Reconfiguration Task
+
+The reconfiguration task will perform the following actions:
+1. Monitor the `wait` condition
+The `wait` condition provided in the DDL will be polled by this task, once the
+condition is met the replicas will be marked as finished and the task will move
+on to finalization.
+
+2. Finalize the reconfiguration
+Finalizing the reconfiguration will occur once the `wait` condition has been
+met. During this phase, the following steps will occur.
+- reconfiguration replica promotion to active
+- cluster configuration update to the new desired state
+- replicas with the previous configuration will be removed
+- replicas for the cluster will be removed from the `mz_pending_cluster_replicas` table
+
+3. Reverting the reconfiguration
+On failure or timeout of the reconfiguration, this task will revert to the state
+prior to the `ALTER CLUSTER` DDL. It will remove reconfigure replicas, and ensure the
+cluster config matches the original version.
+
+
+#### Cluster Sequencer
+
+The following roughly defines the new logic for the cluster sequencer when
+performing an alter cluster on a managed cluster.
+
+`mz_sql::plan::AlterClusterOptions` will have additional fields added to keep track of
+the mechanism being used to perform the alter `AlterClusterMechanism::{Default|Duration|UntilCaughtUp}`.
+
+If we see a non-default mechanism we'll ensure that the cluster is managed
+and pass the mechanism to `sequence_alter_cluster_managed_to_managed`.
+This function will need to be updated to either perform a
+`sequence_alter_cluster_managed_to_managed_disruptive` or
+`sequence_alter_cluster_managed_to_managed_graceful` operation.
+
+The `disruptive` version will perform the current delete-before-create mechanism.
+The `graceful` version will perform the following actions:
+1. Validations
+  1. validate that the cluster is not undergoing a reconfiguration
+  2. validate the cluster is managed
+  3. validate the cluster has no sources and now sinks
+  4. the cluster cannot have a refresh schedule (v1)
+2. Creating the reconfiguration replicas
+3. Transactionally creating reconfiguration replicas and inserting the replicas into
+   the `mz_pending_cluster_replicas` table.
+4. Starting a blocking (or backgrounded) reconfiguration task 
+
+
+#### Recoviner from Environmentd Crash
+Initially, we will recover from an environmentd crash by removing all
+reconfiguration replicas and clearing the `mz_pending_cluster_replicas` table,
```

Comment:
I don't think the replicas need to be explicitly removed. On startup coord announces to compute the desired replica state, and compute complies including removing anything that exists but it was not told about (please verify with the compute team, but I'm like medium confident about this).